pallene/inliner.lua

```diff
@@ -0,0 +1,185 @@
+-- Copyright (c) 2020, The Pallene Developers
+-- Pallene is licensed under the MIT license.
+-- Please refer to the LICENSE and AUTHORS files for details
+-- SPDX-License-Identifier: MIT
+
+local ir = require "pallene.ir"
+local util = require "pallene.util"
+
+local inliner = {}
+
+-- return moves corresponding to each argument as they where at the beginning of the inner scope
+local function args_to_moves(args, cmd, top)
+    local arguments = {}
+    for k, _ in ipairs(args) do
+        -- we can just use the srcs from the call (the arguments) on the rhs of the "move" commands
+        -- at the same time, it is also convenient to adjust the dst (lhs) of the "move"
+        -- to values that will not collide with the outer scope (adding top)
+        -- see ir.lua "Function variables" for more details
+        -- for example,
+        --      x5 = f(x2,x3) -- with srcs = {2,3} and args = {5}
+        --      in a function with a total of 9 variables (top = 9)
+        -- would be translated to
+        --      x(9+1) = x2
+        --      x(9+2) = x3
+        --      ...
+        -- so the new variables will not collide with any of the variables already present in the outer scope
+        -- it also have the nice property of being sequential to top
+        table.insert(arguments, ir.Cmd.Move(cmd.loc, k+top, cmd.srcs[k]))
+    end
+    return ir.Cmd.Seq(arguments)
+end
+
+-- changes the called function (a copy preferably) in place.
+-- each return is turned into a sequence of moves.
+local function return_to_moves(called, cmd)
```

Comment:
In the ir.lua we also had the same problem as you're having here with the shadowing. The convention we're using there is to call the outer command `root_cmd`, so the inner one that we use can be just `cmd`.

Reply:
Prefiro dar o nome melhor (`cmd`) para a veriável de dentro, que é usada mais vezes. Ficar usando `_cmd` tem esse `_` que deixa o código com a cara feia.

Além disso, só colocamos um `_` no começo do nome de uma variável para marcar que é uma variável não usada. A nossa configuração do luacheck trata esses nomes de forma inicial, silenciando o warning de variável não usada caso o nome dela comece com `_`.